File: src/NosAi.Protocol/SessionTranscript.cs
using System.Security.Cryptography;

namespace NosAi.Runtime.Gate1;

/// <summary>Which side of the handshake produced a signature.</summary>
public enum HandshakeRole : byte
{
    /// <summary>The runtime, proving itself to the phone.</summary>
    Server = 0x01,

    /// <summary>The phone, proving itself to the runtime.</summary>
    Client = 0x02
}

/// <summary>
/// What each side actually signs during the Gate 1 handshake.
/// </summary>
/// <remarks>
/// <para>
/// Version 1 had the phone sign the server's raw 32-byte challenge. Three things
/// were wrong with that, and this type fixes all three.
/// </para>
/// <para>
/// <b>The phone was a signing oracle.</b> It signed whatever 32 bytes it was
/// handed, so anything that could get a phone to connect could obtain its
/// signature over a value of its choosing. Signing a hash of a fixed-format
/// transcript instead means a peer would need a preimage to steer the signed
/// value, which it cannot produce.
/// </para>
/// <para>
/// <b>Signatures were not bound to a session.</b> Including both nonces means a
/// signature captured from one session proves nothing in another.
/// </para>
/// <para>
/// <b>Signatures were not bound to a direction.</b> Without the role byte, a
/// signature harvested from the phone could be replayed back as the runtime's
/// proof, and the phone would accept its own signature as evidence it was talking
/// to a genuine runtime.
/// </para>
/// <para>
/// Version 3 (ADR-0009) adds both ephemeral key-agreement public keys to the
/// transcript. That is what authenticates the key exchange: an attacker who
/// substitutes an ephemeral key invalidates the signature carrying it, so the
/// session keys derived from it inherit the identities the RSA keys already
/// proved. No second handshake, no second trust root.
/// </para>
/// </remarks>
public static class SessionTranscript
{
    /// <summary>
    /// Domain separator. Changing it invalidates every existing signature, which
    /// is the point: a transcript from another protocol, or another version of
    /// this one, must never verify here.
    /// </summary>
    private static ReadOnlySpan<byte> Label => "NOSAI-GUARD-HANDSHAKE-V3"u8;

    /// <summary>Required length of each side's nonce.</summary>
    public const int NonceLength = 32;

    /// <summary>
    /// Length of an ephemeral P-256 public key on the wire: an uncompressed
    /// X9.62 point, <c>0x04 || X(32) || Y(32)</c>.
    /// </summary>
    public const int EphemeralKeyLength = 65;

    /// <summary>
    /// Role byte used when deriving keys rather than signing.
    /// </summary>
    /// <remarks>
    /// Not a valid <see cref="HandshakeRole"/>, deliberately: a key-derivation
    /// input can then never collide with a digest either side would put a
    /// signature on.
    /// </remarks>
    private const byte BindingRole = 0x00;

    public static byte[] CreateNonce() => RandomNumberGenerator.GetBytes(NonceLength);

    /// <summary>
    /// The digest the given role signs for this handshake.
    /// </summary>
    /// <remarks>
    /// Every field is included in the same order regardless of role, so the two
    /// sides compute over identical material and only the role byte differs. That
    /// is what keeps the two signatures non-interchangeable.
    /// </remarks>
    public static byte[] Compute(
        HandshakeRole role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
        => Digest((byte)role, clientNonce, serverNonce, clientEphemeral, serverEphemeral);

    /// <summary>
    /// The handshake binding used as HKDF salt when deriving session keys.
    /// </summary>
    /// <remarks>
    /// Identical material to <see cref="Compute"/> with the role byte set to
    /// <see cref="BindingRole"/>, so the derived keys are tied to exactly the
    /// handshake both sides signed. A peer that saw a different nonce or a
    /// different ephemeral key derives different keys and simply cannot decrypt.
    /// </remarks>
    public static byte[] ComputeBinding(
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
        => Digest(BindingRole, clientNonce, serverNonce, clientEphemeral, serverEphemeral);

    /// <summary>
    /// The exact bytes hashed by <see cref="Compute"/>, before the SHA-256.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Exposed for one reason: a private key held in a hardware store cannot sign a
    /// digest that was computed outside it. Android's Keystore signs with
    /// <c>SHA256withRSA</c>, which hashes the message itself, so a keystore-backed
    /// device signs <b>this</b> and the runtime verifies the digest exactly as
    /// before — PKCS#1 v1.5 over SHA-256 of the same bytes either way.
    /// </para>
    /// <para>
    /// The equivalence is pinned by a test rather than asserted here, because a
    /// silent divergence would look like a phone that suddenly fails to
    /// authenticate.
    /// </para>
    /// </remarks>
    public static byte[] Message(
        HandshakeRole role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
        => BuildMessage((byte)role, clientNonce, serverNonce, clientEphemeral, serverEphemeral);

    private static byte[] Digest(
        byte role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
        => SHA256.HashData(BuildMessage(role, clientNonce, serverNonce, clientEphemeral, serverEphemeral));

    private static byte[] BuildMessage(
        byte role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
    {
        if (clientNonce.Length != NonceLength)
            throw new ArgumentException($"Client nonce must be {NonceLength} bytes.", nameof(clientNonce));
        if (serverNonce.Length != NonceLength)
            throw new ArgumentException($"Server nonce must be {NonceLength} bytes.", nameof(serverNonce));
        if (clientEphemeral.Length != EphemeralKeyLength)
            throw new ArgumentException($"Client ephemeral key must be {EphemeralKeyLength} bytes.", nameof(clientEphemeral));
        if (serverEphemeral.Length != EphemeralKeyLength)
            throw new ArgumentException($"Server ephemeral key must be {EphemeralKeyLength} bytes.", nameof(serverEphemeral));

        var buffer = new byte[Label.Length + 3 + (NonceLength * 2) + (EphemeralKeyLength * 2)];
        var offset = 0;

        Label.CopyTo(buffer.AsSpan(offset));
        offset += Label.Length;

        // Separators keep the fields unambiguous: without them a different split
        // of the same bytes could produce the same digest.
        buffer[offset++] = 0x00;
        buffer[offset++] = role;
        buffer[offset++] = 0x00;

        clientNonce.CopyTo(buffer.AsSpan(offset));
        offset += NonceLength;
        serverNonce.CopyTo(buffer.AsSpan(offset));
        offset += NonceLength;
        clientEphemeral.CopyTo(buffer.AsSpan(offset));
        offset += EphemeralKeyLength;
        serverEphemeral.CopyTo(buffer.AsSpan(offset));

        return buffer;
    }

    public static byte[] Sign(
        RSA key,
        HandshakeRole role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral)
        => key.SignHash(
            Compute(role, clientNonce, serverNonce, clientEphemeral, serverEphemeral),
            HashAlgorithmName.SHA256,
            RSASignaturePadding.Pkcs1);

    public static bool Verify(
        RSA key,
        HandshakeRole role,
        ReadOnlySpan<byte> clientNonce,
        ReadOnlySpan<byte> serverNonce,
        ReadOnlySpan<byte> clientEphemeral,
        ReadOnlySpan<byte> serverEphemeral,
        ReadOnlySpan<byte> signature)
    {
        try
        {
            return key.VerifyHash(
                Compute(role, clientNonce, serverNonce, clientEphemeral, serverEphemeral),
                signature,
                HashAlgorithmName.SHA256,
                RSASignaturePadding.Pkcs1);
        }
        catch (CryptographicException)
        {
            return false;
        }
    }
}

File: nosai/network/session_transcript.py
"""What each side signs during the Gate 1 handshake (wire version 3).

Must match `SessionTranscript` in `src/NosAi.Protocol/SessionTranscript.cs`
byte-for-byte: a Python client that hashes a different buffer will be refused
by the runtime, and a C# client that hashes a different buffer will be refused
by a Python test double. The layout is therefore pinned by tests, not restated
in comments.

The digest is SHA-256 of:

    label || 0x00 || role || 0x00 || client_nonce(32) || server_nonce(32)
          || client_ephemeral(65) || server_ephemeral(65)

Signatures are PKCS#1 v1.5 over that pre-hashed digest (RSA SignHash), not over
the raw nonces. Signing the raw challenge would put the phone back in the role
of a signing oracle, which is the hole version 2 exists to close.

Version 3 (ADR-0009) adds both ephemeral P-256 public keys to the material. That
is what authenticates the key agreement the session payload is encrypted under:
substituting an ephemeral key invalidates the signature that carried it, so no
second handshake and no second trust root are needed.
"""
from __future__ import annotations

import hashlib
import os

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.asymmetric.utils import Prehashed

LABEL = b"NOSAI-GUARD-HANDSHAKE-V3"
NONCE_LENGTH = 32

#: Length of an ephemeral P-256 public key on the wire: uncompressed X9.62
#: point, ``0x04 || X(32) || Y(32)``. Matches SessionTranscript.EphemeralKeyLength.
EPHEMERAL_KEY_LENGTH = 65


class HandshakeRole:
    """Matches `NosAi.Runtime.Gate1.HandshakeRole`."""

    SERVER = 0x01
    CLIENT = 0x02


ROLE_SERVER = HandshakeRole.SERVER
ROLE_CLIENT = HandshakeRole.CLIENT

#: Role byte used when deriving keys rather than signing. Not a valid signing
#: role, deliberately: a key-derivation input can then never collide with a
#: digest either side would put a signature on.
ROLE_BINDING = 0x00


def create_nonce() -> bytes:
    return os.urandom(NONCE_LENGTH)


def _digest(
    role: int,
    client_nonce: bytes,
    server_nonce: bytes,
    client_ephemeral: bytes,
    server_ephemeral: bytes,
) -> bytes:
    if len(client_nonce) != NONCE_LENGTH:
        raise ValueError(f"client nonce must be {NONCE_LENGTH} bytes")
    if len(server_nonce) != NONCE_LENGTH:
        raise ValueError(f"server nonce must be {NONCE_LENGTH} bytes")
    if len(client_ephemeral) != EPHEMERAL_KEY_LENGTH:
        raise ValueError(f"client ephemeral key must be {EPHEMERAL_KEY_LENGTH} bytes")
    if len(server_ephemeral) != EPHEMERAL_KEY_LENGTH:
        raise ValueError(f"server ephemeral key must be {EPHEMERAL_KEY_LENGTH} bytes")
    buffer = (
        LABEL
        + b"\x00"
        + bytes((role,))
        + b"\x00"
        + client_nonce
        + server_nonce
        + client_ephemeral
        + server_ephemeral
    )
    return hashlib.sha256(buffer).digest()


def compute(
    role: int,
    client_nonce: bytes,
    server_nonce: bytes,
    client_ephemeral: bytes,
    server_ephemeral: bytes,
) -> bytes:
    """The digest the given role signs for this handshake."""
    if role not in (ROLE_SERVER, ROLE_CLIENT):
        raise ValueError("unknown handshake role")
    return _digest(role, client_nonce, server_nonce, client_ephemeral, server_ephemeral)


def compute_binding(
    client_nonce: bytes,
    server_nonce: bytes,
    client_ephemeral: bytes,
    server_ephemeral: bytes,
) -> bytes:
    """The HKDF salt that ties the session keys to this exact handshake."""
    return _digest(ROLE_BINDING, client_nonce, server_nonce, client_ephemeral, server_ephemeral)


def sign(
    key: rsa.RSAPrivateKey,
    role: int,
    client_nonce: bytes,
    server_nonce: bytes,
    client_ephemeral: bytes,
    server_ephemeral: bytes,
) -> bytes:
    digest = compute(role, client_nonce, server_nonce, client_ephemeral, server_ephemeral)
    return key.sign(digest, padding.PKCS1v15(), Prehashed(hashes.SHA256()))


def verify(
    key: rsa.RSAPublicKey | rsa.RSAPrivateKey,
    role: int,
    client_nonce: bytes,
    server_nonce: bytes,
    client_ephemeral: bytes,
    server_ephemeral: bytes,
    signature: bytes,
) -> bool:
    digest = compute(role, client_nonce, server_nonce, client_ephemeral, server_ephemeral)
    public = key if isinstance(key, rsa.RSAPublicKey) else key.public_key()
    try:
        public.verify(signature, digest, padding.PKCS1v15(), Prehashed(hashes.SHA256()))
    except Exception:
        return False
    return True


__all__ = [
    "EPHEMERAL_KEY_LENGTH",
    "LABEL",
    "NONCE_LENGTH",
    "HandshakeRole",
    "ROLE_BINDING",
    "ROLE_CLIENT",
    "ROLE_SERVER",
    "compute",
    "compute_binding",
    "create_nonce",
    "sign",
    "verify",
]
